README.rst
========================
Python Client for GitLab
========================

Copyright (c) 2022 Jérémie DECOCK (www.jdhp.org)

* Web site: http://www.jdhp.org/software_en.html#gitlablib
* Online documentation: http://gitlablib.readthedocs.org
* Examples: http://gitlablib.readthedocs.org/gallery/

* Notebooks: https://github.com/jdhp/gitlab-lib-notebooks
* Source code: https://github.com/jdhp/gitlab-lib
* Issue tracker: https://github.com/jdhp/gitlab-lib/issues
* Python Client for GitLab on PyPI: https://pypi.org/project/gitlablib
* Python Client for GitLab on Anaconda Cloud: https://anaconda.org/jdhp/gitlablib


Description
===========

An unofficial Python 3 client for the GitLab API

Note:

    This project is still in beta stage, so the API is not finalized yet.


Dependencies
============

*  Python >= 3.0

.. _install:

Installation
============

Gnu/Linux
---------

You can install, upgrade, uninstall Python Client for GitLab with these commands (in a
terminal)::

    pip install --pre gitlablib
    pip install --upgrade gitlablib
    pip uninstall gitlablib

Or, if you have downloaded the Python Client for GitLab source code::

    python3 setup.py install

.. There's also a package for Debian/Ubuntu::
.. 
..     sudo apt-get install gitlablib

Windows
-------

.. Note:
.. 
..     The following installation procedure has been tested to work with Python
..     3.4 under Windows 7.
..     It should also work with recent Windows systems.

You can install, upgrade, uninstall Python Client for GitLab with these commands (in a
`command prompt`_)::

    py -m pip install --pre gitlablib
    py -m pip install --upgrade gitlablib
    py -m pip uninstall gitlablib

Or, if you have downloaded the Python Client for GitLab source code::

    py setup.py install

MacOSX
-------

.. Note:
.. 
..     The following installation procedure has been tested to work with Python
..     3.5 under MacOSX 10.9 (*Mavericks*).
..     It should also work with recent MacOSX systems.

You can install, upgrade, uninstall Python Client for GitLab with these commands (in a
terminal)::

    pip install --pre gitlablib
    pip install --upgrade gitlablib
    pip uninstall gitlablib

Or, if you have downloaded the Python Client for GitLab source code::

    python3 setup.py install


Documentation
=============

* Online documentation: http://gitlablib.readthedocs.org
* API documentation: http://gitlablib.readthedocs.org/en/latest/api.html


Example usage
=============

TODO


Bug reports
===========

To search for bugs or report them, please use the Python Client for GitLab Bug Tracker at:

    https://github.com/jdhp/gitlab-lib/issues


License
=======

This project is provided under the terms and conditions of the `MIT License`_.


.. _MIT License: http://opensource.org/licenses/MIT
.. _command prompt: https://en.wikipedia.org/wiki/Cmd.exe


CHANGES.rst
Changelog
=========

0.1 (2022-05-24)
----------------



init-skeleton.sh
#!/bin/sh

# The MIT License
#
# Copyright (c) 2016 Jérémie DECOCK <jd.jdhp@gmail.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.


# SAFETY TESTS ################################################################

# TODO: assert "$(dirname "$0") == "$(pwd)"
#       ...


# LOAD VARIABLES ##############################################################

. ./meta.sh

echo "AUTHOR_NAME: ${AUTHOR_NAME}"
echo "AUTHOR_EMAIL: ${AUTHOR_EMAIL}"
echo "AUTHOR_WEB_SITE: ${AUTHOR_WEB_SITE}"
echo "COPYRIGHT_NOTICE: ${COPYRIGHT_NOTICE}"
echo "PROJECT_INITIAL_DATE: ${PROJECT_INITIAL_DATE}"
echo "PROJECT_SHORT_DESC: ${PROJECT_SHORT_DESC}"
echo "PROJECT_NAME: ${PROJECT_NAME}"
echo "PYTHON_PACKAGE_NAME: ${PYTHON_PACKAGE_NAME}"
echo "PROJECT_GITHUB_ACCOUNT: ${PROJECT_GITHUB_ACCOUNT}"
echo "PROJECT_GITHUB_REPOSITORY_NAME: ${PROJECT_GITHUB_REPOSITORY_NAME}"
echo "PROJECT_GITHUB_URL ${PROJECT_GITHUB_URL}"
echo "PROJECT_ISSUE_TRACKER_URL: ${PROJECT_ISSUE_TRACKER_URL}"
echo "PROJECT_PYPI_URL: ${PROJECT_PYPI_URL}"
echo "PROJECT_WEB_SITE_URL: ${PROJECT_WEB_SITE_URL}"
echo "PROJECT_ONLINE_DOCUMENTATION_URL: ${PROJECT_ONLINE_DOCUMENTATION_URL}"
echo "PROJECT_ONLINE_API_DOCUMENTATION_URL: ${PROJECT_ONLINE_API_DOCUMENTATION_URL}"


# SETUP GIT ###################################################################

#git remote rename origin skeleton
#git remote add origin git@github.com:${PROJECT_GITHUB_ACCOUNT}/${PROJECT_GITHUB_REPOSITORY_NAME}.git
#git push -u origin master

#git submodule init
#git submodule update


# MAKE SUBSTITUTIONS ##########################################################

sed -i "" \
    -e "s/TODO_AUTHOR_NAME/${AUTHOR_NAME}/g" \
    -e "s/TODO_AUTHOR_EMAIL/${AUTHOR_EMAIL}/g" \
    -e "s TODO_AUTHOR_WEB_SITE ${AUTHOR_WEB_SITE} g" \
    -e "s|TODO_COPYRIGHT_NOTICE|${COPYRIGHT_NOTICE}|g" \
    -e "s|TODO_PROJECT_INITIAL_DATE|${PROJECT_INITIAL_DATE}|g" \
    -e "s|TODO_PROJECT_SHORT_DESC|${PROJECT_SHORT_DESC}|g" \
    -e "s;TODO_PROJECT_NAME;${PROJECT_NAME};g" \
    -e "s/TODO_PYTHON_PACKAGE_NAME/${PYTHON_PACKAGE_NAME}/g" \
    -e "s/TODO_PROJECT_GITHUB_ACCOUNT/${PROJECT_GITHUB_ACCOUNT}/g" \
    -e "s/TODO_PROJECT_GITHUB_REPOSITORY_NAME/${PROJECT_GITHUB_REPOSITORY_NAME}/g" \
    -e "s TODO_PROJECT_GITHUB_URL ${PROJECT_GITHUB_URL} g" \
    -e "s TODO_PROJECT_ISSUE_TRACKER_URL ${PROJECT_ISSUE_TRACKER_URL} g" \
    -e "s TODO_PROJECT_PYPI_URL ${PROJECT_PYPI_URL} g" \
    -e "s TODO_PROJECT_WEB_SITE_URL ${PROJECT_WEB_SITE_URL} g" \
    -e "s TODO_PROJECT_ONLINE_DOCUMENTATION_URL ${PROJECT_ONLINE_DOCUMENTATION_URL} g" \
    -e "s TODO_PROJECT_ONLINE_API_DOCUMENTATION_URL ${PROJECT_ONLINE_API_DOCUMENTATION_URL} g" \
    AUTHORS \
    CHANGES.rst \
    .gitlab-ci.yml \
    .travis.yml \
    environment.yml \
    LICENSE \
    meta.make \
    README.rst \
    setup.cfg \
    docs/api.rst \
    docs/conf.py \
    docs/developer.rst \
    docs/index.rst \
    docs/init_sphinx.sh \
    docs/intro.rst \
    docs/make.bat \
    docs/Makefile \
    TODO_PYTHON_PACKAGE_NAME/__init__.py


# FIX TITLES UNDERLINE LENGTH IN RESTRUCTUREDTEXT FILES #######################

PROJECT_NAME_UNDERLINE=$(echo "${PROJECT_NAME}" | tr '[:print:]' '=')

sed -i "" \
    -e "s/^====$/${PROJECT_NAME_UNDERLINE}/" \
    README.rst

sed -i "" \
    -e "s/^====$/${PROJECT_NAME_UNDERLINE}/" \
    docs/index.rst


# RENAME THE ROOT PACKAGE DIRECTORY ###########################################

mv -v TODO_PYTHON_PACKAGE_NAME "${PYTHON_PACKAGE_NAME}"



gitlablib/gitlab.py
#################################################################
# Install the Python requests library: pip install requests
# http://docs.python-requests.org/en/master/user/quickstart/
#################################################################

import requests
import json
import urllib


class GitLab:

    def __init__(self, api_token, gitlab_host):
        self.api_token = api_token
        self.gitlab_host = gitlab_host


    @property
    def request_header(self):
        header_dict = {
            "PRIVATE-TOKEN": self.api_token
        }
        return header_dict


    def make_issue(self, project_id, title, labels, milestone_id, description):
        # https://docs.gitlab.com/ee/api/issues.html#new-issue
        #
        # dict_keys(['id', 'iid', 'project_id', 'title', 'description', 'state', 'created_at', 'updated_at', 'closed_at', 'closed_by', 'labels', 'milestone', 'assignees', 'author', 'type', 'assignee', 'user_notes_count', 'merge_requests_count', 'upvotes', 'downvotes', 'due_date', 'confidential', 'discussion_locked', 'issue_type', 'web_url', 'time_stats', 'task_completion_status', 'blocking_issues_count', 'has_tasks', 'task_status', '_links', 'references', 'moved_to_id', 'service_desk_reply_to'])

        def post_request(post_url):
            resp = requests.post(post_url, headers=self.request_header)

            if resp.status_code != 201:
                raise Exception("Error:" + resp.text)

            json_dict = json.loads(resp.text)
            return json_dict

        title = urllib.parse.quote(title)              # urlencodé
        labels = urllib.parse.quote(labels)            # urlencodé
        description = urllib.parse.quote(description)  # urlencodé

        gitlab_request_url = f"{self.gitlab_host}/api/v4/projects/{project_id}/issues?title={title}&labels={labels}&milestone_id={milestone_id}&description={description}"
        print(gitlab_request_url)

        resp_dict = post_request(gitlab_request_url)
        
        return resp_dict


    def update_issue(self):
        pass

    
    def gitlab_get_request(self, get_url):
        """
        https://docs.gitlab.com/ee/api/issues.html

        Parameters
        ----------
        get_url : [type]
            [description]

        Returns
        -------
        [type]
            [description]

        Raises
        ------
        Exception
            [description]
        Exception
            [description]
        Exception
            [description]
        """
        resp = requests.get(get_url, headers=self.request_header)

        json_list = json.loads(resp.text)
        if resp.status_code != 200:
            raise Exception("Error:" + resp.text)
        if resp.encoding is not None and resp.encoding.lower() != 'utf-8':
            raise Exception("Encoding error:", resp.encoding)
        if resp.apparent_encoding.lower() != 'utf-8':
            raise Exception("Apparent_encoding error:", resp.apparent_encoding)

        #for issue_dict in json_list:
        #    if ("Ã©" in issue_dict["title"]) or ("â€™" in issue_dict["title"]) or ("Ã‰" in issue_dict["title"]) or ("Ã¢" in issue_dict["title"]): # Ã¨
        #        print("Encoding err:", issue_dict["id"], issue_dict["web_url"])

        return json_list, resp


    def gitlab_fetch_all_issues(self):
        issue_list = []

        url = f"{self.gitlab_host}/api/v4/issues?per_page=100&page=1&scope=all"

        json_list, resp = self.gitlab_get_request(url)
        issue_list.extend(json_list)

        num_pages = int(resp.headers['X-Total-Pages'])

        for page in range(2, num_pages+1):
            #print(".", end="", flush=True)
            print("page {}/{}".format(page, num_pages))

            url = f"{self.gitlab_host}/api/v4/issues?per_page=100&page={page}&scope=all"
            json_list, resp = self.gitlab_get_request(url)

            issue_list.extend(json_list)

        return issue_list


    def gitlab_fetch_selected_issues(self, project_iid_list):
        iid_dict = {}
        for project_id, iid in project_iid_list:
            if project_id not in iid_dict:
                iid_dict[project_id] = []
            iid_dict[project_id].append(iid)

        issue_list = []

        # Group requests per project id
        for project_id, iid_list in iid_dict.items():
            url_filters = "".join([f"&iids[]={iid}" for iid in iid_list])
            url = f"{self.gitlab_host}/api/v4/projects/{project_id}/issues?per_page=100&page=1&scope=all" + url_filters

            json_list, resp = self.gitlab_get_request(url)
            issue_list.extend(json_list)

            num_pages = int(resp.headers['X-Total-Pages'])

            for page in range(2, num_pages+1):
                print(f"page {page}/{num_pages}")

                url = f"{self.gitlab_host}/api/v4/projects/{project_id}/issues?per_page=100&page={page}&scope=all" + url_filters

                print(url)
                json_list, resp = self.gitlab_get_request(url)
                issue_list.extend(json_list)

        return issue_list

gitlablib/__init__.py
"""Python Client for GitLab

An unofficial Python 3 client for the GitLab API

Note:

    This project is in beta stage.

Viewing documentation using IPython
-----------------------------------
To see which functions are available in `gitlablib`, type ``gitlablib.<TAB>`` (where
``<TAB>`` refers to the TAB key), or use ``gitlablib.*get_version*?<ENTER>`` (where
``<ENTER>`` refers to the ENTER key) to narrow down the list.  To view the
docstring for a function, use ``gitlablib.get_version?<ENTER>`` (to view the
docstring) and ``gitlablib.get_version??<ENTER>`` (to view the source code).
"""

import gitlablib.gitlab
from gitlablib.gitlab import GitLab

# PEP0440 compatible formatted version, see:
# https://www.python.org/dev/peps/pep-0440/
#
# Generic release markers:
# X.Y
# X.Y.Z # For bugfix releases  
# 
# Admissible pre-release markers:
# X.YaN # Alpha release
# X.YbN # Beta release         
# X.YrcN # Release Candidate   
# X.Y # Final release
#
# Dev branch marker is: 'X.Y.dev' or 'X.Y.devN' where N is an integer.
# 'X.Y.dev0' is the canonical version of 'X.Y.dev'
#
__version__ = '0.1.dev1'

def get_version():
    return __version__

__all__ = ['TODO']


meta.sh
AUTHOR_NAME="Jérémie DECOCK"
AUTHOR_EMAIL="jd.jdhp@gmail.com"
AUTHOR_WEB_SITE="www.jdhp.org"

COPYRIGHT_NOTICE="Copyright (c) 2022 ${AUTHOR_NAME} (${AUTHOR_WEB_SITE})"

PROJECT_INITIAL_DATE="2022-05-24"
PROJECT_SHORT_DESC="An unofficial Python 3 client for the GitLab API"

PROJECT_NAME="Python Client for GitLab"
PYTHON_PACKAGE_NAME="gitlablib"                 # TODO e.g. pyax12 (all in lower case and without special character)

PROJECT_GITHUB_ACCOUNT="jdhp"              # TODO (GitLab)
#PROJECT_GITHUB_ACCOUNT="jeremiedecock"     # TODO (GitHub)
PROJECT_GITHUB_REPOSITORY_NAME="gitlab-lib"

PROJECT_PYPI_URL="https://pypi.org/project/${PYTHON_PACKAGE_NAME}"
PROJECT_WEB_SITE_URL="http://www.jdhp.org/software_en.html#${PYTHON_PACKAGE_NAME}"

# GITLAB ##############################

#PROJECT_GITHUB_URL="https://gitlab.com/${PROJECT_GITHUB_ACCOUNT}/${PROJECT_GITHUB_REPOSITORY_NAME}"
#PROJECT_ISSUE_TRACKER_URL="https://gitlab.com/${PROJECT_GITHUB_ACCOUNT}/${PROJECT_GITHUB_REPOSITORY_NAME}/issues"

#PROJECT_ONLINE_DOCUMENTATION_URL="https://${PROJECT_GITHUB_ACCOUNT}.gitlab.io/${PYTHON_PACKAGE_NAME}"
#PROJECT_ONLINE_API_DOCUMENTATION_URL="https://${PROJECT_GITHUB_ACCOUNT}.gitlab.io/${PYTHON_PACKAGE_NAME}/api.html"

# GITHUB ##############################

PROJECT_GITHUB_URL="https://github.com/${PROJECT_GITHUB_ACCOUNT}/${PROJECT_GITHUB_REPOSITORY_NAME}"
PROJECT_ISSUE_TRACKER_URL="https://github.com/${PROJECT_GITHUB_ACCOUNT}/${PROJECT_GITHUB_REPOSITORY_NAME}/issues"

PROJECT_ONLINE_DOCUMENTATION_URL="http://${PYTHON_PACKAGE_NAME}.readthedocs.org"
PROJECT_ONLINE_API_DOCUMENTATION_URL="http://${PYTHON_PACKAGE_NAME}.readthedocs.org/en/latest/api.html"
